### PR TITLE
Fixes #1075

### LINKED ISSUE #1075: Mixing normalization and concatenation
This problem was reported by @jamescheney in #1067. It's unrelated to deduplication, but caused by a bug in the mixing normaliser:

```
links> query {for (x <-- factorials) [(a=x.i),(a=x.f)]};
[(a = 1), (a = 1), (a = 2), (a = 2), (a = 3), (a = 6)] : [(a:Int)]
links> query mixing {for (x <-- factorials) [(a=x.i),(a=x.f)]};
***: Fatal error : body 
links> query nested {for (x <-- factorials) [(a=x.i),(a=x.f)]};
[(a = 1), (a = 2), (a = 3), (a = 1), (a = 2), (a = 6)] : [(a:Int)]
links> query delat {for (x <-- factorials) [(a=x.i),(a=x.f)]};
***: Fatal error : body
```

The mixing normaliser does not split comprehensions along concatenation correctly. It then returns an expression in the shape `U{ {a}++{b} | x <-- factorials }` which is not in normal form, so the conversion to SQL fails. The culprit is one between `MixingQuery.reduce_for_source/reduce_for_body` (or perhaps their combination).

### PR DESCRIPTION
**Pull request by wricciot**
This is straightforward: the mixing normaliser (`mixingQuery.ml`) has an issue with `For` when its body is a `Concat`: the comprehension would normally be split along concatenation, but this does not happen. 

The issue is caused by the changes made to normalisation to allow it to consider normal forms where comprehension generators can be complex queries: in order for these changes to work, we must add a recursive call to `MixingQuery.reduce_for_body` (which until now was non-recursive).

The old normaliser doesn't show this bug, for reasons not entirely clear (but related to the simpler structure of homogeneous bag normal forms). However it's now advisable that the code responsible for normalising comprehensions should be refactored to make its rationale more understandable (without a clear rationale, every further change is at risk of creating new bugs).

### CODE REVIEW

File: core/query/mixingQuery.ml [hunk 1]
```diff
@@ -208,10 +208,11 @@ let rec reduce_where_then (c, t) =
   | _ ->
     Q.If (c, t, Q.Concat [])
 
-let reduce_for_body (gs, os, body) =
+let rec reduce_for_body (gs, os, body) =
+  let rb = reduce_for_body in
   match body with
     (* | Concat []                 -> body *)
```
Comment (jamescheney) [inline core/query/mixingQuery.ml:214]: Is it worth removing this commented-out line now that Concat is handled uniformly by a new case?

**Comment by jamescheney** [comment]
If it helps, here is the result of manually running the queries from `mixing.links` with sqlite3:

```
Finishing process MAIN
 _     _ __   _ _  __  ___
/ |   | |  \ | | |/ / / ._\
| |   | | , \| |   /  \  \
| |___| | |\ \ | |\ \ _\  \
|_____|_|_| \__|_| \_|____/
Welcome to Links version 0.9.6 (Burghmuirhead)
links> var db = database "links";
debug: found loadable database driver
- /home/jcheney/src/tmp/links-issue1075/_build/install/default/lib/links-sqlite3/links_sqlite3.cmxs
- /home/jcheney/src/tmp/links-issue1075/_build/install/default/share/links-sqlite3/links_sqlite3_dependencies.json

registering driver for sqlite3
Finishing process MAIN
db = (database sqlite3:links : Database
links> var bagtable = table "bagtable" with (i : Int) from db;
Finishing process MAIN
bagtable = (table bagtable) : TableHandle((i:Int),(i:Int),(i:Int))
links> query mixing { distinct(bagtable) }
...... ;
Query.eval time: 44
Running query: 
select distinct (t2340."i") as "i"
from (select distinct *
from "bagtable" as t2340) as t2340
select distinct (t2340."i") as "i"
from (select distinct *
from "bagtable" as t2340) as t2340
Finishing process MAIN
[(i = 0)] : [(i:Int)]
links> query mixing { dedup(for (x <-- bagtable) [x]) };
Query.eval time: 35
Running query: 
select distinct (t2347."i") as "i"
from (select distinct *
from "bagtable" as t2347) as t2347
select distinct (t2347."i") as "i"
from (select distinct *
from "bagtable" as t2347) as t2347
Finishing process MAIN
[(i = 0)] : [(i:Int)]
links> query mixing { for (x <- distinct(bagtable)) [x] };
Query.eval time: 87
Running query: 
select (t2355."i") as "i"
from (select distinct (t2354."i") as "i"
from (select distinct *
from "bagtable" as t2354) as t2354) as t2355
select (t2355."i") as "i"
from (select distinct (t2354."i") as "i"
from (select distinct *
from "bagtable" as t2354) as t2354) as t2355
Finishing process MAIN
[(i = 0)] : [(i:Int)]
links> query mixing { for (x <-- bagtable) [(a=x.i),(a=x.i)] };
Query.eval time: 39
Running query: 
(select (t2361."i") as "a"
from "bagtable" as t2361)
union all
(select (t2361."i") as "a"
from "bagtable" as t2361)
(select (t2361."i") as "a"
from "bagtable" as t2361)
union all
(select (t2361."i") as "a"
from "bagtable" as t2361)
***: Error: Sqlite3.Error("Sqlite3.prepare: near \"(\": syntax error") 
links> 
```
Trying the query in sqlite3 directly, it seems that 
```
select (t2361."i") as "a" from "bagtable" as t2361
```
works but 
```
(select (t2361."i") as "a" from "bagtable" as t2361)
```
is a syntax error, likewise
```
select (t2361."i") as "a" from "bagtable" as t2361 union all select (t2361."i") as "a" from "bagtable" as t2361
```
works but
```
(select (t2361."i") as "a" from "bagtable" as t2361) union all (select (t2361."i") as "a" from "bagtable" as t2361)
```
is a syntax error.  I guess the parentheses are not strictly necessary in SQL (at least as long as all of the operations are "union all") since  Postgres and Mysql must allow a more relaxed syntax.  How strange!

**Comment by wricciot** [comment]
> ```
> select (t2361."i") as "a" from "bagtable" as t2361 union all select (t2361."i") as "a" from "bagtable" as t2361
> ```
> 
> works but
> 
> ```
> (select (t2361."i") as "a" from "bagtable" as t2361) union all (select (t2361."i") as "a" from "bagtable" as t2361)
> ```
> 
> is a syntax error. I guess the parentheses are not strictly necessary in SQL (at least as long as all of the operations are "union all") since Postgres and Mysql must allow a more relaxed syntax. How strange!

So much for "one extra parenthesis never hurt anyone"!

Thanks!

**Comment by jamescheney** [comment]
It seems the SQL standard does not require that parentheses are allowed here, see: https://stackoverflow.com/questions/4653124/what-does-the-sql-standard-say-about-parentheses-in-sql-union-except-intersect-s

Perhaps interestingly if you mix UNION and UNION ALL at the same level then Q UNION Q UNION ALL Q and Q UNION ALL Q UNION Q yield different results.  It seems the first one gets evaluated as (Q UNION Q) UNION ALL Q (i.e. you get dedup(Q) UNION ALL Q) while the second gets evaluated as (Q UNION ALL Q) UNION Q (you get dedup(Q) since the deduplicating union happens last).  

One can emulate the parentheses using subqueries however, e.g. `Q1 UNION (Q2 UNION ALL Q3)` can be written as `Q1 UNION SELECT * FROM (Q1 UNION Q2)`.  So if there are places where we might generate mixes of union and union all (or other set operators) then it might make sense to program defensively here to avoid relying on syntax extensions not required by SQL.

**Comment by wricciot** [comment]
Sigh, welcome to the bureau of complication of simple affairs. Thanks!

**Comment by jamescheney** [comment]
I suppose one way to fix this could be to add a method `parenthesize` to the database class/interface that abstracts over a safe way to parenthesize SQL queries; MySQL and Postgres could implement this by just adding parentheses while sqlite3 could do it the more conservative/defensive way.  This seems a little ugly though, especially if we want pretty-printing to play nicely with it.  So maybe better to just do the horrible `select * from (...)` thing everywhere.  Now that I think of it though, this only seems necessary for arguments to a `Union` that are not `Select`s, since our `Union` construct is homogeneous (that is, with a single `Union` you can only do a bunch of "union all"s, or a bunch of "union"s, not a mix).

**Comment by jamescheney** [comment]
Merging in the fix to #1090 should hopefully make CI happy now.

**Comment by dhil** [comment]
Is this one good to go in now?

**Comment by wricciot** [comment]
It is ready as far as I'm concerned.

**Pull request merged**